File: tools/gui/src/domain/schemas/promotion.ts
import { Schema } from 'effect'

export const PromotionConflictSchema = Schema.Struct({
  engagement_id: Schema.String,
  enterprise_id: Schema.String,
  artifact_type: Schema.String,
  engagement_name: Schema.String,
  enterprise_name: Schema.String,
  engagement_fields: Schema.Record({ key: Schema.String, value: Schema.Unknown }),
  enterprise_fields: Schema.Record({ key: Schema.String, value: Schema.Unknown }),
})
export type PromotionConflict = typeof PromotionConflictSchema.Type

export const PromotionDocumentConflictSchema = Schema.Struct({
  engagement_id: Schema.String,
  enterprise_id: Schema.String,
  doc_type: Schema.String,
  engagement_title: Schema.String,
  enterprise_title: Schema.String,
})
export type PromotionDocumentConflict = typeof PromotionDocumentConflictSchema.Type

export const PromotionDiagramConflictSchema = Schema.Struct({
  engagement_id: Schema.String,
  enterprise_id: Schema.String,
  diagram_type: Schema.String,
  engagement_name: Schema.String,
  enterprise_name: Schema.String,
})
export type PromotionDiagramConflict = typeof PromotionDiagramConflictSchema.Type

export const PromotionGroupMappingEntrySchema = Schema.Struct({
  engagement_slug: Schema.String,
  engagement_group_id: Schema.String,
  match_status: Schema.Literal('matched_by_id', 'conflict', 'new'),
  enterprise_slug: Schema.String,
  enterprise_group_id: Schema.NullOr(Schema.String),
})
export type PromotionGroupMappingEntry = typeof PromotionGroupMappingEntrySchema.Type

export const StructuralClosureEntitySchema = Schema.Struct({
  artifact_id: Schema.String,
  name: Schema.String,
  artifact_type: Schema.String,
})
export type StructuralClosureEntity = typeof StructuralClosureEntitySchema.Type

/** One selected junction/grouping whose meaning-carrying entities are missing from the
 * promotion selection — the GUI offers a one-action "include the missing entities" flow
 * from exactly this data. */
export const StructuralClosureRequirementSchema = Schema.Struct({
  entity_id: Schema.String,
  entity_name: Schema.String,
  kind: Schema.Literal('junction', 'grouping'),
  missing: Schema.Array(StructuralClosureEntitySchema),
})
export type StructuralClosureRequirement = typeof StructuralClosureRequirementSchema.Type

/** A viewpoint a promoted diagram or matrix is pinned to. `enterprise_version` null means it is not
 *  there at all — the case that forces a choice between promoting it alongside and repinning. */
export const PromotionViewpointDependencySchema = Schema.Struct({
  target_id: Schema.String,
  target_kind: Schema.String,
  slug: Schema.String,
  pinned_version: Schema.String,
  status: Schema.String,
  enterprise_version: Schema.NullOr(Schema.String),
})
export type PromotionViewpointDependency = typeof PromotionViewpointDependencySchema.Type

/** An artifact the selection references but does not include, and what needs it. */
export const PromotionMissingDependencySchema = Schema.Struct({
  artifact_id: Schema.String,
  name: Schema.String,
  record_type: Schema.String,
  required_by: Schema.String,
  kind: Schema.String,
})
export type PromotionMissingDependency = typeof PromotionMissingDependencySchema.Type

export const EnterpriseGroupOptionSchema = Schema.Struct({
  slug: Schema.String,
  id: Schema.String,
  name: Schema.String,
})

/**
 * What a promotion would do, and every question it raises before it may proceed.
 *
 * Nothing here is optional now. Four of these lists were, and two were absent outright —
 * `viewpoint_dependencies` and `missing_dependencies`, the two that can make a promotion produce a
 * broken enterprise repository. An empty list and an uncomputed one have to be distinguishable, and
 * only one of them is representable when the key may be missing.
 */
export const PromotionPlanSchema = Schema.Struct({
  entity_id: Schema.String,
  entities_to_add: Schema.Array(Schema.String),
  conflicts: Schema.Array(PromotionConflictSchema),
  connection_ids: Schema.Array(Schema.String),
  already_in_enterprise: Schema.Array(Schema.String),
  warnings: Schema.Array(Schema.String),
  documents_to_add: Schema.Array(Schema.String),
  diagrams_to_add: Schema.Array(Schema.String),
  doc_conflicts: Schema.Array(PromotionDocumentConflictSchema),
  diagram_conflicts: Schema.Array(PromotionDiagramConflictSchema),
  schema_errors: Schema.Array(Schema.String),
  structural_closure: Schema.Array(StructuralClosureRequirementSchema),
  group_mapping: Schema.Array(PromotionGroupMappingEntrySchema),
  available_enterprise_groups: Schema.Array(EnterpriseGroupOptionSchema),
  viewpoint_dependencies: Schema.Array(PromotionViewpointDependencySchema),
  missing_dependencies: Schema.Array(PromotionMissingDependencySchema),
})
export type PromotionPlan = typeof PromotionPlanSchema.Type

export const PromotionResultSchema = Schema.Struct({
  dry_run: Schema.Boolean,
  executed: Schema.Boolean,
  copied_files: Schema.Array(Schema.String),
  updated_files: Schema.Array(Schema.String),
  verification_errors: Schema.Array(Schema.String),
  // The one to read after a failure: the promotion runs in a git worktree transaction, so saying the
  // enterprise repository was restored is what stops an operator cleaning up by hand.
  rolled_back: Schema.Boolean,
  warnings: Schema.Array(Schema.String),
})
export type PromotionResult = typeof PromotionResultSchema.Type

File: src/infrastructure/rest/contracts/promotion.py
"""Response contracts for promotion: the plan, and the result of carrying it out.

Promotion copies engagement artifacts into the enterprise repository, so the *plan* is the thing a
person reads before agreeing to it — every list on it is a question they have to answer. That is why
none of them are optional here: a plan that omits a key when it is empty makes "nothing conflicts" and
"conflicts not computed" the same reply, and the client's own decoder had four of them optional and two
missing outright.

Derived from ``promote_to_enterprise.plan_promotion`` and the two branches of ``promotion_execute``.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict


class _Closed(BaseModel):
    model_config = ConfigDict(extra="forbid")


class PromotionEntityConflict(_Closed):
    """An entity that exists in both repositories, with both readings side by side.

    ``engagement_fields`` and ``enterprise_fields`` are the two versions' frontmatter, published so the
    resolution UI can show a field-by-field diff and offer a merge — which is the whole point of
    surfacing a conflict rather than refusing the promotion.
    """

    engagement_id: str
    enterprise_id: str
    artifact_type: str
    engagement_name: str
    enterprise_name: str
    engagement_fields: dict[str, Any]
    enterprise_fields: dict[str, Any]


class PromotionDocumentConflict(_Closed):
    """A document that exists in both repositories."""

    engagement_id: str
    enterprise_id: str
    doc_type: str
    engagement_title: str
    enterprise_title: str


class PromotionDiagramConflict(_Closed):
    """A diagram that exists in both repositories."""

    engagement_id: str
    enterprise_id: str
    diagram_type: str
    engagement_name: str
    enterprise_name: str


class PromotionGroupMappingEntry(_Closed):
    """Where one engagement group lands in the enterprise repository.

    ``match_status`` is the decision: matched by id, a conflicting slug, or new. ``enterprise_group_id``
    is null for ``new`` — there is nothing to point at yet.
    """

    engagement_slug: str
    engagement_group_id: str
    match_status: Literal["matched_by_id", "conflict", "new"]
    enterprise_slug: str
    enterprise_group_id: str | None


class EnterpriseGroupOption(_Closed):
    """One enterprise group a mapping may be redirected to."""

    slug: str
    id: str
    name: str


class StructuralClosureEntity(_Closed):
    """One entity missing from the selection."""

    artifact_id: str
    name: str
    artifact_type: str


class StructuralClosureRequirement(_Closed):
    """A selected junction or grouping whose meaning-carrying entities are not in the selection.

    A junction promoted without what it joins is a structure that means nothing at the far end, so the
    entities are named rather than the promotion being refused: the GUI offers "include the missing
    entities" from exactly this data.
    """

    entity_id: str
    entity_name: str
    kind: Literal["junction", "grouping"]
    missing: list[StructuralClosureEntity]


class PromotionViewpointDependency(_Closed):
    """A viewpoint a promoted diagram or matrix is pinned to, and where it stands enterprise-side.

    ``enterprise_version`` is null when the viewpoint is not there at all — which is the case that
    forces a choice between promoting it alongside and repinning, and the reason this list exists.
    """

    target_id: str
    target_kind: str
    slug: str
    pinned_version: str
    status: str
    enterprise_version: str | None


class PromotionMissingDependency(_Closed):
    """An artifact the selection references but does not include, and what needs it."""

    artifact_id: str
    name: str
    record_type: str
    required_by: str
    kind: str


class PromotionPlanResponse(_Closed):
    """What a promotion would do, and every question it raises before it may proceed.

    Nothing here is optional. Four of these lists were optional in the client's decoder and two were
    absent from it — ``viewpoint_dependencies`` and ``missing_dependencies``, which are the two that
    can make a promotion produce a broken enterprise repository. An empty list and an uncomputed one
    have to be distinguishable, and only one of them is representable when the key may be missing.
    """

    entity_id: str
    entities_to_add: list[str]
    conflicts: list[PromotionEntityConflict]
    connection_ids: list[str]
    already_in_enterprise: list[str]
    documents_to_add: list[str]
    diagrams_to_add: list[str]
    doc_conflicts: list[PromotionDocumentConflict]
    diagram_conflicts: list[PromotionDiagramConflict]
    warnings: list[str]
    schema_errors: list[str]
    structural_closure: list[StructuralClosureRequirement]
    group_mapping: list[PromotionGroupMappingEntry]
    available_enterprise_groups: list[EnterpriseGroupOption]
    viewpoint_dependencies: list[PromotionViewpointDependency]
    missing_dependencies: list[PromotionMissingDependency]


class PromotionResultResponse(_Closed):
    """What carrying out the plan actually did.

    ``dry_run`` and ``executed`` are both stated because they answer different questions: a dry run
    never executes, but an execution can also fail to, and a caller distinguishing "we did not try"
    from "we tried and stopped" needs both.

    ``rolled_back`` is the one to read after a failure. The promotion runs in a git worktree
    transaction, so a verification error leaves the enterprise repository as it was — and saying so is
    what stops an operator from going to clean up by hand.

    ``warnings`` carries the plan's, which the dry-run branch does not currently populate; the key is
    always present so a reader never has to tell an absent list from an empty one.
    """

    dry_run: bool
    executed: bool
    copied_files: list[str]
    updated_files: list[str]
    verification_errors: list[str]
    rolled_back: bool
    warnings: list[str] = []
